### PR TITLE
[Manta] Refactor the code base w.r.t. manta-crypto, pallet-manta-asset and pallet-manta-pay

### PR DESCRIPTION
**Pull request by zhenfeizhang**
* remove dependency on `manta-client`
* update code to depend on `manta-crypto`, `pallet-manta-asset` and `pallet-manta-pay`

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -58,8 +58,9 @@ sp-core = { version = "3.0.0", package = "sp-core" }
 substrate-subxt-client = { version = "0.7.0", path = "client", optional = true }
 substrate-subxt-proc-macro = { version = "0.15.0", path = "proc-macro" }
 
-pallet-manta-pay = { git = "https://github.com/Manta-Network/pallet-manta-pay", branch = "calamari", default-features = false, features = [ "std" ] }
-manta-client = { git = "https://github.com/Manta-Network/manta-client", branch = "calamari" }
+pallet-manta-pay = { branch = "refactor", git = "https://github.com/Manta-Network/pallet-manta-pay", default-features = false, features = [ "std" ] }
+pallet-manta-asset = { branch = "integrate-token-from-pallet-manta-pay", git = "https://github.com/Manta-Network/pallet-manta-asset", default-features = false, features = [ "std" ] }
+manta-crypto = { branch = "refactoring-pallet-manta-pay's-crypto-code", git = "https://github.com/Manta-Network/manta-crypto", default-features = false, features = [ "std" ] }
```
Comment (zhenfeizhang) [inline Cargo.toml:63]: those two dependencies will be updated once the branches are merged.

**Review by gabcbrown — COMMENTED** [review]
A couple of minor questions + comments, otherwise looks good.

File: examples/manta_mint_asset.rs [hunk 2]
```diff
@@ -16,18 +16,18 @@
 // You should have received a copy of the GNU General Public License
 // along with Manta-subxt.  If not, see <http://www.gnu.org/licenses/>.
 
+use manta_crypto::{
+    CommitmentParam,
+    MantaSerDes,
+    COMMIT_PARAM,
```
[the remark this answers is not in the document]
Comment (zhenfeizhang) [reply examples/manta_mint_asset.rs:22]: CommitmentParam is an alias. It wraps some parameters from the underlying arkworks library.
COMMIT_PARAM is a constant, which is essentially the deserialized form of an instance of CommitmentParam.

File: examples/manta_transfer_private_asset.rs [hunk 3]
```diff
@@ -95,28 +97,13 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
-    let tree_1 = LedgerMerkleTree::new(hash_param.clone(), &shard_1.list).unwrap();
-    let index_over_tree = shard_1
-        .list
-        .iter()
-        .position(|r| *r == sender_coin_1.0.cm_bytes)
-        .unwrap();
-    let path_1 = tree_1
-        .generate_proof(index_over_tree, &sender_coin_1.0.cm_bytes)
-        .unwrap();
+    let shard_1 = shards.shard[sender_coin_1.commitment[0] as usize].clone();
```
[the remark this answers is not in the document]
Comment (zhenfeizhang) [reply examples/manta_transfer_private_asset.rs:100]: nice catch

File: src/manta_misc.rs [hunk 4]
```diff
@@ -55,24 +54,21 @@ pub async fn manta_mint_token_helper<R: RngCore + CryptoRng>(
...
-
-    let mut mint_data = [0u8; 96];
-    mint_data[0..32].copy_from_slice(&coin.0.cm_bytes);
-    mint_data[32..64].copy_from_slice(&coin.1.k);
-    mint_data[64..96].copy_from_slice(&coin.1.s);
-
+    let asset = MantaAsset::sample(&commit_param, &sk, &amount, rng);
+    let mint_data = generate_mint_payload(&asset);
+    let mut payload = [0u8; 104];
+    mint_data.serialize(payload.as_mut());
```
[the remark this answers is not in the document]
Comment (zhenfeizhang) [reply src/manta_misc.rs:64]: yeah. good point. I think it was imported from `pallet_manta_pay`. I need to change the interface there.

File: examples/manta_transfer_private_asset.rs [hunk 5]
```diff
@@ -126,46 +113,26 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
...
 
     // ====================================
     // 3. build the circuit
     // ====================================
     let circuit = TransferCircuit {
         // param
-        commit_param,
-        hash_param,
+        commit_param: commit_param.clone(),
+        hash_param: hash_param.clone(),
```
[the remark this answers is not in the document]
Comment (zhenfeizhang) [reply examples/manta_transfer_private_asset.rs:127]: I think it is the later case

> Or are they just references for how one might use this code

File: src/manta_misc.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhenfeizhang) [reply src/manta_misc.rs]: fixed with 
https://github.com/Manta-Network/pallet-manta-pay/pull/39/commits/4eb3c87b590c71394f643345cf4b03c390fe652e and https://github.com/Manta-Network/manta-subxt/pull/8/commits/319fd558d53795e40a3d3248abb951035d96077e

**Review by gabcbrown — APPROVED** [review]

**Comment by Dengjianping** [comment]
LGTM. But I think we should remove these `unwrap`s in another pr, better error handling makes our code more safe.

**Comment by zhenfeizhang** [comment]
Agreed. Creating an Issue for it.

**Pull request merged**